domain/secret/state/state.go

```diff
@@ -357,30 +415,280 @@ func (st State) createSecret(
 	return nil
 }
 
-func updateSecretMetadataFromParams(p domainsecret.UpsertSecretParams, md *secretMetadata) {
-	md.Description = p.Description
-	md.AutoPrune = p.AutoPrune
-	if p.RotatePolicy != nil {
-		md.RotatePolicyID = int(*p.RotatePolicy)
+// createSecret creates the records needed to store secret data, excluding secret owner records.
+func (st State) updateSecret(
+	ctx context.Context, tx *sqlair.TX, uri *coresecrets.URI,
+	secret domainsecret.UpsertSecretParams,
+) error {
+	// We need the latest revision so far, plus owner info for the secret,
+	// so we may as well also include existing metadata as well so simplify
+	// the update statement needed.
+	existingSecretQuery := fmt.Sprintf(`
+WITH rev AS
+    (SELECT  MAX(revision) AS latest_revision
+    FROM     secret_revision
+    WHERE    secret_id = $M.id)
+SELECT 
+     (sm.secret_id,
+     version,
+     description,
+     auto_prune,
+     rp.policy,
+     rev.latest_revision) AS (&secretInfo.*),
+     (so.owner_kind,
+     so.owner_id,
+     so.label) AS (&secretOwner.*)
+FROM secret_metadata sm, rev
+       LEFT JOIN secret_rotate_policy rp ON rp.id = sm.rotate_policy_id
+       LEFT JOIN (
+          SELECT '%s' AS owner_kind, '' AS owner_id, label, secret_id
+          FROM   secret_model_owner so
+       UNION
+          SELECT '%s' AS owner_kind, application.uuid AS owner_id, label, secret_id
+          FROM   secret_application_owner so
+          JOIN   application
+          WHERE  application.uuid = so.application_uuid
+       UNION
+          SELECT '%s' AS owner_kind, unit_uuid AS owner_id, label, secret_id
+          FROM   secret_unit_owner so
+          JOIN   unit
+          WHERE  unit.uuid = so.unit_uuid
+       ) so ON so.secret_id = sm.secret_id
+WHERE sm.secret_id = $M.id
+`, coresecrets.ModelOwner, coresecrets.ApplicationOwner, coresecrets.UnitOwner)
+
+	existingSecretStmt, err := st.Prepare(existingSecretQuery, sqlair.M{}, secretInfo{}, secretOwner{})
```

Comment:
When we were speaking with Gustavo about scalar inputs, he suggested that we should avoid over-reliance on maps. Certainly, where we repeat any access patterns, it's going to be more efficient to use small types for parameterisation. I'd be inclined to do that for secret IDs.

Reply:
I've added some small bespoke types